### PR TITLE
Implement `IndexUId` as `(String, Ulid)`

### PR DESCRIPTION
**Pull request by guilload**
Draft proposal.

### CODE REVIEW

File: quickwit/quickwit-proto/protos/quickwit/metastore.proto [hunk 1]
```diff
@@ -23,6 +23,11 @@ package quickwit.metastore;
 
 import "quickwit/ingest.proto";
 
+message IndexUid {
```
Comment (guilload) [inline quickwit/quickwit-proto/protos/quickwit/metastore.proto:26]: Mapping this message to `quickwit_proto::types::IndexUid` using the extern path feature, we are never exposed to that message.

File: quickwit/quickwit-proto/protos/quickwit/metastore.proto [hunk 2]
```diff
@@ -216,8 +221,8 @@ message DeleteTask {
 }
 
 message DeleteQuery {
-  // Index ID.
-  string index_uid = 1;
+  // Index UID.
+  IndexUid index_uid = 1;
```
Comment (guilload) [inline quickwit/quickwit-proto/protos/quickwit/metastore.proto:225]: Now, we have a typed `IndexUid` in protobuf messages. Yay!
Now, we have to deal with an option. Nay!

File: quickwit/quickwit-proto/src/metastore/mod.rs [hunk 3]
```diff
@@ -221,6 +223,14 @@ impl CloseShardsFailure {
     }
 }
 
+impl DeleteQuery {
+    pub fn index_uid(&self) -> &IndexUid {
```
Comment (guilload) [inline quickwit/quickwit-proto/src/metastore/mod.rs:227]: One solution to get a `&IndexUid`.

File: quickwit/quickwit-proto/src/types/index_uid.rs [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+}
+
+impl<'de> Deserialize<'de> for IndexUid {
+    fn deserialize<D: serde::Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
+        unimplemented!()
+    }
+}
+
+/// Encodes an index UID into a protobuf message composed of one string and two u64 integers.
+impl ::prost::Message for IndexUid {
```
Comment (guilload) [inline quickwit/quickwit-proto/src/types/index_uid.rs:91]: The trick. If this is too complex for your taste, we can always go with `(String, String)`.

File: quickwit/quickwit-proto/src/types/index_uid.rs [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+    }
+
+    fn clear(&mut self) {
+        self.index_id.clear();
+        self.incarnation_id = Ulid::nil();
+    }
+}
+
+#[macro_export]
+macro_rules! index_uid {
```
Comment (guilload) [inline quickwit/quickwit-proto/src/types/index_uid.rs:178]: One way to move the `index_uid` field from an option without moving the other fields of the protobuf message.

File: quickwit/quickwit-proto/src/types/mod.rs [hunk 6]
```diff
@@ -47,83 +49,10 @@ pub fn queue_id(index_uid: &str, source_id: &str, shard_id: u64) -> QueueId {
...
-    }
-
-    pub fn as_str(&self) -> &str {
-        &self.0
-    }
-
-    pub fn from_parts(index_id: impl Into<String>, incarnation_id: impl Into<String>) -> Self {
-        let incarnation_id = incarnation_id.into();
-        let index_id = index_id.into();
-        if incarnation_id.is_empty() {
```
Comment (guilload) [inline quickwit/quickwit-proto/src/types/mod.rs:81]: This is needed for backward compatibility. We should no longer need this as we're dropping backward compatibility with 0.4 and 0.5 in the next release.

File: quickwit/quickwit-proto/protos/quickwit/metastore.proto (hunk 2 above)
Comment (trinity-1686a) [reply quickwit/quickwit-proto/protos/quickwit/metastore.proto:225]: I think that's a breaking change and would technically require to change the field number. I don't know how much we care about that though.

**Review by trinity-1686a — COMMENTED** [review]
I really like the idea of having proper typing as soon as deserialization end. On the other hand, I don't like the manual impl of `prost::Message`, but I don't think that's avoidable.

Do you want me to take it from there?

File: quickwit/quickwit-proto/src/types/index_uid.rs [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+        match tag {
+            1u32 => {
+                let value = &mut self.index_id;
+                ::prost::encoding::string::merge(wire_type, value, buf, ctx).map_err(|mut error| {
+                    error.push(STRUCT_NAME, "index_id");
+                    error
+                })
+            }
+            2u32 => {
+                let (mut ulid_high, ulid_low) = self.incarnation_id.into();
```
[the remark this answers is not in the document]
Comment (trinity-1686a) [reply quickwit/quickwit-proto/src/types/index_uid.rs:128]: I doubt we would be able to measure the difference. Our serialization does a fair bit of allocation, for instance for the 1st field of that same struct. The variance from just that would probably make measurements too noisy to figure out if anything changed

File: quickwit/quickwit-proto/src/types/index_uid.rs (hunk 5 above)
Comment (guilload) [reply quickwit/quickwit-proto/src/types/index_uid.rs:178]: Trinity suggests using the getter and cloning if an owned object is required. The call to `clone()` will hopefully be optimized away. I agree :)

File: quickwit/quickwit-proto/src/types/index_uid.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (guilload) [reply quickwit/quickwit-proto/src/types/index_uid.rs:128]: The `From` conversions in the `ulid` crate do exactly that.

**Pull request closed without merging**